File: Realtime Raytracing/python/13 image based lighting/finished/engine.py
from config import *
import scene
import sphere
import node
import materials

class Engine:
    """
        Responsible for drawing scenes
    """

    def __init__(self, width, height):
        """
            Initialize a flat raytracing context
            
                Parameters:
                    width (int): width of screen
                    height (int): height of screen
        """
        self.screenWidth = width
        self.screenHeight = height

        self.targetFrameRate = 60
        self.frameRateMargin = 10

        #general OpenGL configuration
        self.shader = self.createShader("shaders/frameBufferVertex.txt",
                                        "shaders/frameBufferFragment.txt")
        
        self.rayTracerShader = self.createComputeShader("shaders/rayTracer.txt")
        
        glUseProgram(self.shader)
        
        self.createQuad()
        self.createLODChain()
        self.createColorBuffers()
        self.createResourceMemory()
        self.skyBoxMaterial = materials.CubeMapMaterial("gfx/sky")
        glUseProgram(self.rayTracerShader)
        glUniform1i(glGetUniformLocation(self.rayTracerShader, "sky_cube"), 4)
    
    def createQuad(self):
        # x, y, z, s, t
        self.vertices = np.array(
            ( 1.0,  1.0, 0.0, 1.0, 1.0, #top-right
             -1.0,  1.0, 0.0, 0.0, 1.0, #top-left
             -1.0, -1.0, 0.0, 0.0, 0.0, #bottom-left
             -1.0, -1.0, 0.0, 0.0, 0.0, #bottom-left
              1.0, -1.0, 0.0, 1.0, 0.0, #bottom-right
              1.0,  1.0, 0.0, 1.0, 1.0), #top-right
             dtype=np.float32
        )

        self.vertex_count = 6

        self.vao = glGenVertexArrays(1)
        glBindVertexArray(self.vao)
        self.vbo = glGenBuffers(1)
        glBindBuffer(GL_ARRAY_BUFFER, self.vbo)
        glBufferData(GL_ARRAY_BUFFER, self.vertices.nbytes, self.vertices, GL_STATIC_DRAW)

        glEnableVertexAttribArray(0)
        glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 20, ctypes.c_void_p(0))
        glEnableVertexAttribArray(1)
        glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 20, ctypes.c_void_p(12))
    
    def createLODChain(self):

        self.resolutions = [(self.screenWidth, self.screenHeight)]

        width,height = (self.screenWidth,self.screenHeight)
        while width > 2 and height > 2:
            width = int(width / 1.25)
            height = int(height / 1.25)
            self.resolutions.append((width, height))
        
        self.resolutionLevel = len(self.resolutions) - 1
        
        self.screenWidth,self.screenHeight = self.resolutions[self.resolutionLevel]

    def createColorBuffers(self):

        self.colorBuffers = []

        for resolution in self.resolutions:

            width,height = resolution

            newColorBuffer = glGenTextures(1)
            self.colorBuffers.append(newColorBuffer)
            glActiveTexture(GL_TEXTURE0)
            glBindTexture(GL_TEXTURE_2D, newColorBuffer)

            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)

        
            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA32F, width, height, 0, GL_RGBA, GL_FLOAT, None)
        
        self.colorBuffer = self.colorBuffers[self.resolutionLevel]
    
    def createResourceMemory(self):

        """
            allocate storage for up to 1024 objects (why not?)
        """

        # sphere: (cx cy cz r) (r g b roughness)
        self.objectData = np.zeros(8 * 1024, dtype = np.float32)

        self.objectDataTexture = glGenTextures(1)
        glActiveTexture(GL_TEXTURE1)
        glBindTexture(GL_TEXTURE_2D, self.objectDataTexture)

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)
    
        glTexImage2D(GL_TEXTURE_2D,0,GL_RGBA32F, 2, 1024,0,GL_RGBA,GL_FLOAT,bytes(self.objectData))

        # node: (x_min, y_min, z_min, x_max) (y_max, z_max, index hit_link) (miss_link, offset, count, _)
        self.nodeData = np.zeros(12 * 1024, dtype = np.float32)

        self.nodeTexture = glGenTextures(1)
        glActiveTexture(GL_TEXTURE2)
        glBindTexture(GL_TEXTURE_2D, self.nodeTexture)

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)
    
        glTexImage2D(GL_TEXTURE_2D,0,GL_RGBA32F,3,1024,0,GL_RGBA,GL_FLOAT,bytes(self.nodeData))

        # sphere_lookup: (index _ _ _)
        self.sphereLookupData = np.zeros(4 * 1024, dtype = np.float32)

        self.sphereLookupTexture = glGenTextures(1)
        glActiveTexture(GL_TEXTURE3)
        glBindTexture(GL_TEXTURE_2D, self.sphereLookupTexture)

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)
    
        glTexImage2D(GL_TEXTURE_2D,0,GL_RGBA32F,1,1024,0,GL_RGBA,GL_FLOAT,bytes(self.sphereLookupData))

    def createShader(self, vertexFilepath, fragmentFilepath):
        """
            Read source code, compile and link shaders.
            Returns the compiled and linked program.
        """

        with open(vertexFilepath,'r') as f:
            vertex_src = f.readlines()

        with open(fragmentFilepath,'r') as f:
            fragment_src = f.readlines()
        
        shader = compileProgram(compileShader(vertex_src, GL_VERTEX_SHADER),
                                compileShader(fragment_src, GL_FRAGMENT_SHADER))
        
        return shader
    
    def createComputeShader(self, filepath):
        """
            Read source code, compile and link shaders.
            Returns the compiled and linked program.
        """

        with open(filepath,'r') as f:
            compute_src = f.readlines()
        
        shader = compileProgram(compileShader(compute_src, GL_COMPUTE_SHADER))
        
        return shader

    def recordSphere(self, i, _sphere: sphere.Sphere):

        # sphere: (cx cy cz r) (r g b roughness)

        self.objectData[8*i]     = _sphere.center[0]
        self.objectData[8*i + 1] = _sphere.center[1]
        self.objectData[8*i + 2] = _sphere.center[2]

        self.objectData[8*i + 3] = _sphere.radius

        self.objectData[8*i + 4] = _sphere.color[0]
        self.objectData[8*i + 5] = _sphere.color[1]
        self.objectData[8*i + 6] = _sphere.color[2]

        self.objectData[8*i + 7] = _sphere.roughness
    
    def recordNode(self, i, _node: node.Node):

        # node: (x_min, y_min, z_min, x_max) (y_max, z_max, index hit_link) (miss_link, offset, count, _)

        self.nodeData[12*i]      = _node.min_corner[0]
        self.nodeData[12*i + 1]  = _node.min_corner[1]
        self.nodeData[12*i + 2]  = _node.min_corner[2]

        self.nodeData[12*i + 3]  = _node.max_corner[0]
        self.nodeData[12*i + 4]  = _node.max_corner[1]
        self.nodeData[12*i + 5]  = _node.max_corner[2]

        self.nodeData[12*i + 6]  = _node.index

        self.nodeData[12*i + 7]  = _node.hit_link
        self.nodeData[12*i + 8]  = _node.miss_link

        self.nodeData[12*i + 9]  = _node.first_sphere_index
        self.nodeData[12*i + 10] = _node.sphere_count
    
    def recordSphereIndex(self, i, index):

        # index: (index, _, _, _)

        self.sphereLookupData[4*i] = index
    
    def updateScene(self, scene: scene.Scene):

        scene.outDated = False

        glUseProgram(self.rayTracerShader)

        glUniform1f(glGetUniformLocation(self.rayTracerShader, "sphereCount"), len(scene.spheres))

        for i,_sphere in enumerate(scene.spheres):
            self.recordSphere(i, _sphere)
        
        glActiveTexture(GL_TEXTURE1)
        glBindTexture(GL_TEXTURE_2D, self.objectDataTexture)
        glTexImage2D(GL_TEXTURE_2D,0,GL_RGBA32F,2,1024,0,GL_RGBA,GL_FLOAT,bytes(self.objectData))

        glUniform1f(glGetUniformLocation(self.rayTracerShader, "nodeCount"), scene.nodes_used)

        for i in range(scene.nodes_used):
            _node = scene.nodes[i]
            if _node.min_corner is not None:
                self.recordNode(i, _node)

        glActiveTexture(GL_TEXTURE2)
        glBindTexture(GL_TEXTURE_2D, self.nodeTexture)
        glTexImage2D(GL_TEXTURE_2D,0,GL_RGBA32F,3,1024,0,GL_RGBA,GL_FLOAT,bytes(self.nodeData))

        for i,_index in enumerate(scene.sphere_ids):
            self.recordSphereIndex(i, _index)

        glActiveTexture(GL_TEXTURE3)
        glBindTexture(GL_TEXTURE_2D, self.sphereLookupTexture)
        glTexImage2D(GL_TEXTURE_2D,0,GL_RGBA32F,1,1024,0,GL_RGBA,GL_FLOAT,bytes(self.sphereLookupData))

    def prepareScene(self, scene: scene.Scene):
        """
            Send scene data to the shader.
        """

        glUseProgram(self.rayTracerShader)

        glUniform3fv(glGetUniformLocation(self.rayTracerShader, "viewer.position"), 1, scene.camera.position)
        glUniform3fv(glGetUniformLocation(self.rayTracerShader, "viewer.forwards"), 1, scene.camera.forwards)
        glUniform3fv(glGetUniformLocation(self.rayTracerShader, "viewer.right"), 1, scene.camera.right)
        glUniform3fv(glGetUniformLocation(self.rayTracerShader, "viewer.up"), 1, scene.camera.up)

        if scene.outDated:
            self.updateScene(scene)
        
        glActiveTexture(GL_TEXTURE1)
        glBindImageTexture(1, self.objectDataTexture, 0, GL_FALSE, 0, GL_READ_ONLY, GL_RGBA32F)

        glActiveTexture(GL_TEXTURE2)
        glBindImageTexture(2, self.nodeTexture, 0, GL_FALSE, 0, GL_READ_ONLY, GL_RGBA32F)

        glActiveTexture(GL_TEXTURE3)
        glBindImageTexture(3, self.sphereLookupTexture, 0, GL_FALSE, 0, GL_READ_ONLY, GL_RGBA32F)

        self.skyBoxMaterial.use()
        
    def renderScene(self, scene: scene.Scene):
        """
            Draw all objects in the scene
        """
        #start = time.time()
        glUseProgram(self.rayTracerShader)

        self.prepareScene(scene)

        glActiveTexture(GL_TEXTURE0)
        glBindImageTexture(0, self.colorBuffer, 0, GL_FALSE, 0, GL_WRITE_ONLY, GL_RGBA32F)
        
        glDispatchCompute(self.screenWidth, self.screenHeight, 1)
  
        # make sure writing to image has finished before read
        glMemoryBarrier(GL_SHADER_IMAGE_ACCESS_BARRIER_BIT)
        glBindImageTexture(0, 0, 0, GL_FALSE, 0, GL_WRITE_ONLY, GL_RGBA32F)
        self.drawScreen()
        #finish = time.time()
        #print(f"render took {(finish - start) * 1000} milliseconds.")

    def drawScreen(self):
        glUseProgram(self.shader)
        glBindFramebuffer(GL_FRAMEBUFFER, 0)
        glActiveTexture(GL_TEXTURE0)
        glBindTexture(GL_TEXTURE_2D, self.colorBuffer)
        glBindVertexArray(self.vao)
        glDrawArrays(GL_TRIANGLES, 0, self.vertex_count)
        pg.display.flip()
    
    def adaptResolution(self, frameRate):

        if frameRate > self.targetFrameRate + self.frameRateMargin and self.resolutionLevel > 0:
            #increase resolution
            self.resolutionLevel -= 1
        elif frameRate < self.targetFrameRate - self.frameRateMargin and self.resolutionLevel < len(self.resolutions) - 1:
            #reduce resolution
            self.resolutionLevel += 1
        
        self.screenWidth,self.screenHeight = self.resolutions[self.resolutionLevel]
        self.colorBuffer = self.colorBuffers[self.resolutionLevel]
    
    def destroy(self):
        """
            Free any allocated memory
        """
        glUseProgram(self.rayTracerShader)
        glMemoryBarrier(GL_ALL_BARRIER_BITS)
        glDeleteProgram(self.rayTracerShader)
        glDeleteVertexArrays(1, (self.vao,))
        glDeleteBuffers(1, (self.vbo,))
        glDeleteTextures(1, (self.colorBuffer,))
        glDeleteProgram(self.shader)

File: Projects/Bounding Volume Hierarchies/sparse/view/engine.cpp
#include "engine.h"
#include "../model/node.h"

Engine::Engine(int width, int height) {

	shader = util::load_shader("shaders/vertex.txt", "shaders/fragment.txt");
	glUseProgram(shader);

	glClearColor(0.0f, 0.0f, 0.0f, 1.0f);

	this->width = width;
	this->height = height;
	screenMesh = new QuadModel;
	
	targetFramerate = 30;
	framerateMargin = 5;
	sampleCount = 4;
	bounceCount = 4;

	create_LOD_chain();
	create_color_buffers();

}

Engine::~Engine() {
	delete screenMesh;
	glDeleteTextures(colorBuffers.size(), colorBuffers.data());
	glDeleteProgram(shader);
}

void Engine::create_LOD_chain() {

	unsigned int tempWidth = width;
	unsigned int tempHeight = height;
	resolutions.push_back({ tempWidth, tempHeight });
	

	while (tempWidth > 2 && tempHeight > 2) {
		tempWidth = static_cast<unsigned int>(tempWidth / 1.25);
		tempHeight = static_cast<unsigned int>(tempHeight / 1.25);
		resolutions.push_back({ tempWidth, tempHeight });
	}

	resolutionLevel = static_cast<unsigned int>(resolutions.size()) - 1;
	//resolutionLevel = 0;
	width = resolutions[resolutionLevel].width;
	height = resolutions[resolutionLevel].height;

}

void Engine::create_color_buffers() {

	for (FrameSize frame : resolutions) {

		unsigned int tempWidth = frame.width;
		unsigned int tempHeight = frame.height;

		unsigned int newColorBuffer = 0;
		glGenTextures(1, &newColorBuffer);
		glActiveTexture(GL_TEXTURE0);
		glBindTexture(GL_TEXTURE_2D, newColorBuffer);

		glTextureParameteri(newColorBuffer, GL_TEXTURE_WRAP_S, GL_REPEAT);
		glTextureParameteri(newColorBuffer, GL_TEXTURE_WRAP_T, GL_REPEAT);
		glTextureParameteri(newColorBuffer, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
		glTextureParameteri(newColorBuffer, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
		glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, NULL);
		
		colorBuffers.push_back(newColorBuffer);

		std::vector <uint32_t> newColorBufferMemory;
		for (int y = 0; y < tempHeight; ++y) {
			for (int x = 0; x < tempWidth; ++x) {
				newColorBufferMemory.push_back(0);
			}
		}
		colorBufferMemories.push_back(newColorBufferMemory);
	}

	colorBuffer = colorBuffers[resolutionLevel];
	colorBufferMemory = colorBufferMemories[resolutionLevel];

}

void Engine::pset(size_t x, size_t y, glm::vec3 color) {

	uint8_t r = std::max(0,std::min(255,(int)(255 * glm::sqrt(color.x))));
	uint8_t g = std::max(0, std::min(255,(int)(255 * glm::sqrt(color.y))));
	uint8_t b = std::max(0, std::min(255,(int)(255 * glm::sqrt(color.z))));
	colorBufferMemory[x + (size_t)width * y] = r + (g << 8) + (b << 16) + (255 << 24);

}

glm::vec3 Engine::ray_color(Ray& ray, Scene* scene, int depth) {

	hit_record rec;

	if (depth <= 0) {
		return glm::vec3(0.0f);
	}

	bool hit_something = false;
	float t_min = 0.00001f;
	float t_max = 9999;

	//Traverse scene's BVH
	std::queue<Node*> nodes;
	nodes.push(scene->root);
	while (!nodes.empty()) {
		Node* node = nodes.front();
		nodes.pop();

		if (!node) {
			continue;
		}

		if (node->volume->hit(ray)) {

			if (node->isInternal) {
				for (Node* child : node->children) {
					if (child) {
						nodes.push(child);
					}
				}
			}
			else {
				for (Sphere* sphere : node->spheres) {
					if (sphere->hit(ray, t_min, t_max, rec)) {
						hit_something = true;
						t_max = rec.t;
					}
				}

			}
		}

	}

	if (hit_something) {
		Ray newRay(rec.point, glm::normalize(rec.normal + 0.9f * random_in_unit_sphere()));
		return 0.5f * rec.color * ray_color(newRay, scene, depth - 1);
	}

	glm::vec3 unit_direction{ glm::normalize(ray.direction) };
	float t = 0.5f * (unit_direction.z + 1.0f);
	return (1.0f - t) * glm::vec3(1.0f, 1.0f, 1.0f) + t * glm::vec3(0.5f, 0.7f, 1.0f);
}

void Engine::render(Scene* scene) {

	//auto start = std::chrono::high_resolution_clock::now();

	for (int y = 0; y < height; ++y) {
		//std::cout << y << " / " << height << std::endl;
		for (int x = 0; x < width; ++x) {
			glm::vec3 color = glm::vec3(0.0f);
			for (int sample = 0; sample < sampleCount; ++sample) {
				//float horizontalCoefficient = ((float(x) * 2 - width) / width);
				//float verticalCoefficient = ((float(y) * 2 - height) / width);
				float horizontalCoefficient = 2.0f * (x + random_float()) / (width)-1.0f;
				float verticalCoefficient = 2.0f * (y + random_float()) / (width)-1.0f;
				Ray ray(scene->player->position, glm::normalize(scene->player->forwards + horizontalCoefficient * scene->player->right + verticalCoefficient * scene->player->up));
				color = color + (1.0f / sampleCount) * ray_color(ray, scene, bounceCount);
			}
			pset(x, y, color);
		}
	}

	//auto stop = std::chrono::high_resolution_clock::now();
	//auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start);
	//std::cout << "This render took " << duration.count() << " ms." << std::endl;
}

void Engine::draw_screen() {

	glUseProgram(shader);

	glActiveTexture(GL_TEXTURE0);
	glBindTexture(GL_TEXTURE_2D, colorBuffer);

	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, colorBufferMemory.data());

	glBindVertexArray(screenMesh->VAO);
	glDrawArrays(GL_TRIANGLES, 0, 6);
	glFlush();

}

void Engine::adapt_framerate(int framerate) {

	if (framerate > targetFramerate + framerateMargin && resolutionLevel > 0) {
		resolutionLevel -= 1;
	}
	else if (framerate < targetFramerate - framerateMargin && resolutionLevel < resolutions.size() - 1) {
		resolutionLevel += 1;
	}

	width = resolutions[resolutionLevel].width;
	height = resolutions[resolutionLevel].height;

	colorBuffer = colorBuffers[resolutionLevel];
	colorBufferMemory = colorBufferMemories[resolutionLevel];
}